diff --git a/build/parser.js b/build/parser.js
index f293154d4601e786270652c4e7a40f7a724a422b..590ed6d58cb6b2cded1d334c6c429cd2626869b0 100644
--- a/build/parser.js
+++ b/build/parser.js
@@ -110,7 +110,7 @@ export default class Parser {
          *       attireBlock
          *   )
          */
-        if (closingTokens.length && this.peekToken.Type === Tokens.SLASH) {
+        if (closingTokens.includes(Tokens.RPAREN) && this.peekToken.Type === Tokens.SLASH) {
             const propertyType = [];
             while (!closingTokens.includes(this.curToken.Type)) {
                 propertyType.push(...this.parsePropertyTypes());
@@ -123,6 +123,9 @@ export default class Parser {
                     this.nextToken();
                 }
             }
+            if (this.curToken.Type === Tokens.RPAREN) {
+                this.nextToken();
+            }
             if (groupName) {
                 const variable = {
                     Type: 'variable',
@@ -163,6 +166,24 @@ export default class Parser {
             return [prop];
         }
         while (!closingTokens.includes(this.curToken.Type)) {
+            /**
+             * check if we have a group choice instead of an assignment
+             */
+            if (this.curToken.Type === Tokens.SLASH && this.peekToken.Type === Tokens.SLASH) {
+                if (valuesOrProperties.length === 0) {
+                    throw this.parserError('Unexpected group choice operator "//" at start of group');
+                }
+
+                if (!isChoice) {
+                    const last = valuesOrProperties.pop();
+                    valuesOrProperties.push([last]);
+                    isChoice = true;
+                }
+                this.nextToken();
+                this.nextToken();
+                continue;
+            }
+
             const propertyType = [];
             const comments = [];
             let isUnwrapped = false;
@@ -197,13 +218,24 @@ export default class Parser {
                 this.curToken.Literal === Tokens.LBRACK ||
                 this.curToken.Literal === Tokens.LPAREN) {
                 const innerGroup = this.parseAssignmentValue();
-                valuesOrProperties.push({
+                const prop = {
                     HasCut: false,
                     Occurrence: occurrence,
                     Name: '',
                     Type: innerGroup,
                     Comments: []
-                });
+                };
+
+                if (isChoice) {
+                    valuesOrProperties[valuesOrProperties.length - 1].push(prop);
+                } else {
+                    valuesOrProperties.push(prop);
+                }
+
+                if (this.curToken.Type === Tokens.COMMA) {
+                    this.nextToken();
+                    isChoice = false;
+                }
                 continue;
             }
             /**
@@ -485,10 +517,6 @@ export default class Parser {
     openSegment() {
         if (this.curToken.Type === Tokens.LBRACE) {
             this.nextToken();
-            if (this.peekToken.Type === Tokens.LPAREN) {
-                this.nextToken();
-                return [Tokens.RPAREN, Tokens.RBRACE];
-            }
             return [Tokens.RBRACE];
         }
         else if (this.curToken.Type === Tokens.LPAREN) {
@@ -549,6 +577,9 @@ export default class Parser {
                         Unwrapped: isUnwrapped
                     };
                 }
+                else if (this.curToken.Literal === Tokens.LBRACE) {
+                    type = this.parseAssignmentValue();
+                }
                 else if (this.curToken.Type === Tokens.IDENT) {
                     type = {
                         Type: 'group',
